http_server_proxy.go

```diff
@@ -0,0 +1,197 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"strconv"
+)
+
+// HTTPServerProxyStoreMode definition
+type HTTPServerProxyStoreMode string
+
+const (
+	// HTTPServerProxyStoreModePassthrough mode
+	HTTPServerProxyStoreModePassthrough HTTPServerProxyStoreMode = "passthru"
+)
+
+// HTTPServerProxyStoreErrorResponseBody definition
+type HTTPServerProxyStoreErrorResponseBody struct {
+	Error string `json:"error"`
+}
+
+// HTTPServerProxyStoreRequestData definition
+type HTTPServerProxyStoreRequestData struct {
+	Method  string      `json:"method"`
+	Path    string      `json:"path"`
+	Headers http.Header `json:"header"`
+	Query   url.Values  `json:"query"`
+	Body    []byte      `json:"body"`
+}
+
+// HTTPServerProxyStoreResponseData definition
+type HTTPServerProxyStoreResponseData struct {
+	StatusCode int                                  `json:"statuscode"`
+	Headers    http.Header                          `json:"header"`
+	Body       HTTPServerProxyStoreResponseDataBody `json:"body"`
+}
+
+// HTTPServerProxyStoreResponseDataBody definition
+type HTTPServerProxyStoreResponseDataBody struct {
+	Offset int `json:"offset"`
+}
+
+// HTTPServerProxyStoreDataEntry definition
+type HTTPServerProxyStoreDataEntry struct {
+	Offset   int                              `json:"offset"`
+	Request  HTTPServerProxyStoreRequestData  `json:"request"`
+	Response HTTPServerProxyStoreResponseData `json:"response"`
+}
+
+// HTTPServerProxyStore definition
+type HTTPServerProxyStore struct {
+	Name string                   `json:"name"`
+	Mode HTTPServerProxyStoreMode `json:"mode"`
+	Data []HTTPServerProxyStoreDataEntry
+}
+
+// HTTPServerProxy definition
+type HTTPServerProxy map[string]HTTPServerProxyStore
+
+// NewProxy allocates a new http proxy from its configuration
+func NewProxy(cfg HTTPServerProxy) *HTTPServerProxy {
+	proxy := new(HTTPServerProxy)
+	*proxy = cfg
+	return proxy
+}
+
+// Listen for the proxy store/retrieve
+func (proxy *HTTPServerProxy) Listen(mux *http.ServeMux, prefix string) {
+	for p, s := range *proxy {
+		store := HTTPServerProxyStore{Name: p, Mode: s.Mode, Data: []HTTPServerProxyStoreDataEntry{}}
+		mux.HandleFunc(fmt.Sprintf("%swrite/%s", prefix, p), store.write)
+		mux.HandleFunc(fmt.Sprintf("%sread/%s", prefix, p), store.read)
+	}
+}
+
+// write stores incoming request data
+func (store *HTTPServerProxyStore) write(w http.ResponseWriter, r *http.Request) {
+
+	reqData := HTTPServerProxyStoreRequestData{
+		Method:  r.Method,
+		Path:    r.URL.RequestURI(),
+		Headers: r.Header,
+		Query:   r.URL.Query(),
+	}
+
+	offset := len(store.Data)
+
+	resData := HTTPServerProxyStoreResponseData{
+		StatusCode: http.StatusOK,
+		Headers: http.Header{
+			"Content-Type": []string{"application/json"},
+		},
+		Body: HTTPServerProxyStoreResponseDataBody{
+			Offset: offset,
+		},
+	}
```

Comment:
I dont get the whole thing here. Please rework it so it returns the body direct.

use shorter variables / function names, use packages.

i am not checking this code further.